SO-2881: SCT RF2 export request changes

apeteri opened the pull request.

apeteri committed:
SO-2881: Add enum for maintaining organization type

This controls whether "INT", an ISO3166 country code or an empty prefix
should appear in the CountryNamespace element of each exported file.

apeteri committed:
SO-2881: Create sub-section in SnomedCoreConfiguration for RF2 defaults

Initial values are the following:

- The maintainer is set to "INT"
- No namespace restriction
- Reference sets are exported to a single file per reference set type

apeteri committed:
SO-2881: Add enum for defining the RF2 reference set export type

apeteri committed:
SO-2881: Change extension namespace ID property name in...

...SnomedExportDefaultConfiguration

apeteri committed:
SO-2881: Add initial version of new RF2 exporting request

apeteri committed:
SO-2881: Move new RF2 import-related files to their own package

apeteri committed:
SO-2881: Update new RF2 importer classes after move

apeteri committed:
SO-2881: Update SnomedExportRestService after signature change

apeteri committed:
SO-2881: Fix issues related to RF2 core export

- Wrap requests in a RevisionIndexReadRequests as well
- Retrieve all code system version branches
- Use transient effective time for file names if pre-release content is
included
- Do not reuse the byte buffer holding CRLF characters
- Initialize scrolling search requests with a timeout first, use
scrollId in subsequent invocations

apeteri committed:
SO-2881: Update new RF2 export classes

- Add support for reference sets
- Invoke reference set exporters differently, depending on the requested
export layout
- Add language reference set exporter with sub-query for language code

apeteri committed:
Merge remote-tracking branch 'origin/6.x' into issue/SO-2881_RF2_export_changes

apeteri committed:
SO-2881: Add language code to filename in Rf2LanguageRefSetExporter

apeteri committed:
SO-2881: Skip language reference set exporting if the IDs are empty

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java:
```diff
@@ -15,273 +15,777 @@
+				});
+				break;
+			case FULL:
+				// Full export: nothing to remove, all versions should be exported
```
I would add/keep the ability to filter by effective time in case of Delta and Full export types.

apeteri replied:
If you filter by effective time, it is no longer a full export. On the other hand, a delta export with the effective time range left untouched will give you a full export 😄

cmark replied:
You're right! 👍

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java:
```diff
@@ -15,273 +15,777 @@
+		this.nrcCountryCode = nrcCountryCode;
+	}
+
+	void setNamespaceId(final String namespaceId) {
```
Is this the ID of the namespace concept? Wouldn't it be better to use the namespace ID directly? LanguageSetting is going to be deprecated soon, so we either check all FSNs or outsource namespace ID extraction to the clients.

apeteri replied:
Moving computation to the client.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java:
```diff
@@ -15,273 +15,777 @@
+				? "BETA"
+						: "PRODUCTION";
+
+		final Path releaseDirectory = exportDirectory.resolve(String.format("SNOMEDCT_RF2_%s_%sT120000Z", releaseStatus, latestEffectiveTime));
```
⚠️ Hardcoded hours!

apeteri replied:
The release root directory name is computed from the transient effective date in this case, which does not include time-level information.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java:
```diff
@@ -15,273 +15,777 @@
+	private Set<String> getLanguageCodes(final RepositoryContext context, final String branch) {
+		final Set<String> languageCodes = newHashSet();
+
+		// TODO: there should be an easier way than trying all possible language codes...
```
Extracting all possible `languageCode` field values from descriptions index?

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/exporter/Rf2Exporter.java:
```diff
@@ -0,0 +1,184 @@
+	
+	private static final String CR_LF = "\r\n";
+
+	private static final int BATCH_SIZE = 1000;
```
Increase batch size to the `5000`-`10000` range.

apeteri committed:
SO-2881: Move country-namespace element and namespace filter...

...computations to the client

apeteri committed:
SO-2881: Increase export batch size to 10000 items

apeteri committed:
SO-2881: Populate missing properties in SnomedExportRestService

apeteri committed:
SO-2881: Maintain backwards compatibility on SnomedExportRestService

If a DELTA export is requested without an effective time range, assume
that pre-release components should be exported only.

apeteri committed:
SO-2881: Multiple changes to RF2 exporter implementations

- Keep the effective time that is used in the filename separate from
what goes into the RF2 rows
- Allow skipping effective time filtering altogether by passing in
Long.MIN_VALUE

apeteri committed:
SO-2881: Multiple changes to SnomedRf2ExportRequest

- Fix effective time filtering and set of visited branches for snapshot
exports
- Pass transient effective time and "latest" effective time to
RF2Exporters
- Fix cases where a null value is set of componentTypes, modules and
refSets (under these conditions, all possible values should be used
without explicitly enumerating them)

apeteri committed:
SO-2881: Make code system version candidate list writable

cmark reviewed (approved):
LGTM!
